### PR TITLE
Refactor LB GC

### PR DESCRIPTION
**Pull request by spencerhance**

### CODE REVIEW

File: pkg/loadbalancers/l7s.go [hunk 1]
```diff
@@ -82,74 +82,91 @@ func (l *L7s) Delete(name string, version meta.Version, scope meta.KeyType) erro
...
+		}
+	}
+
+	urlMaps, err := l.List(meta.GlobalKey(""), meta.VersionGA)
 	if err != nil {
-		return err
+		return fmt.Errorf("error listing global LBs: %v", err)
+	}
+
+	if errors := l.gc(urlMaps, knownLoadBalancers, features.NewResourceVersions()); errors != nil {
```
Comment (bowei) [inline pkg/loadbalancers/l7s.go]: We should probably alias `var GAResourceVersions = NewResourceVersions()` in the features package to make this more understandable.

File: pkg/loadbalancers/l7s.go [hunk 2]
```diff
@@ -82,74 +82,91 @@ func (l *L7s) Delete(name string, version meta.Version, scope meta.KeyType) erro
...
-	} else {
-		urlMaps, err = composite.ListUrlMaps(l.cloud, meta.GlobalKey(""), meta.VersionGA)
-	}
+// List returns a list of urlMaps (the top level LB resource) that belong to the cluster
+func (l *L7s) List(key *meta.Key, version meta.Version) ([]*composite.UrlMap, error) {
+	var result []*composite.UrlMap
+	urlMaps, err := composite.ListUrlMaps(l.cloud, key, version)
 	if err != nil {
-		return nil, nil, err
+		return nil, fmt.Errorf("error listing url maps: %v", err)
```
[the remark this answers is not in the document]
Comment (spencerhance) [reply pkg/loadbalancers/l7s.go]: Done

File: pkg/loadbalancers/l7s.go [hunk 3]
```diff
@@ -82,74 +82,91 @@ func (l *L7s) Delete(name string, version meta.Version, scope meta.KeyType) erro
...
+		if err != nil {
+			return fmt.Errorf("error listing regional LBs: %v", err)
+		}
+
+		if err := l.gc(urlMaps, knownLoadBalancers, features.L7ILBVersions()); err != nil {
+			return fmt.Errorf("error gc-ing regional LBs: %v", err)
+		}
+	}
+
+	urlMaps, err := l.List(meta.GlobalKey(""), meta.VersionGA)
```
[the remark this answers is not in the document]
Comment (spencerhance) [reply pkg/loadbalancers/l7s.go]: Done

File: pkg/loadbalancers/l7s.go [hunk 4]
```diff
@@ -82,74 +82,91 @@ func (l *L7s) Delete(name string, version meta.Version, scope meta.KeyType) erro
...
+	var errors []error
+
 	// Delete unknown loadbalancers
-	for i, name := range pool {
-		if knownLoadBalancers.Has(name) {
+	for _, um := range urlMaps {
+		nameParts := l.namer.ParseName(um.Name)
+		l7Name := l.namer.LoadBalancerFromLbName(nameParts.LbName)
+
+		if knownLoadBalancers.Has(l7Name) {
```
[the remark this answers is not in the document]
Comment (spencerhance) [reply pkg/loadbalancers/l7s.go]: Done

File: pkg/loadbalancers/l7s.go (hunk 1 above)
Comment (spencerhance) [reply pkg/loadbalancers/l7s.go]: Done

File: pkg/loadbalancers/features/features.go [hunk 5]
```diff
@@ -28,6 +28,8 @@ const (
 	FeatureL7ILB = "L7ILB"
 )
 
+var GAResourceVersions = NewResourceVersions()
```
Comment (bowei) [inline pkg/loadbalancers/features/features.go]: Doc string

**Comment by bowei** [comment]
/lgtm
/approve

you need to fix golint

**Pull request merged**